pyccel/parser/semantic.py

```diff
@@ -2375,34 +2374,51 @@ def _visit_FunctionDef(self, expr, **settings):
-                    d_var['is_kwonly'] = a.is_kwonly
-                    dtype = d_var.pop('datatype')
-                    if d_var['rank']>0:
-                        d_var['cls_base'] = NumpyArrayClass
-
-                    # this is needed for the static case
-
                     additional_args = []
-                    if isinstance(a, ValuedArgument):
+                    if isinstance(ah, FunctionAddress):
+                        d_var = {}
+                        d_var['is_argument'] = True
+                        d_var['is_pointer'] = True
+                        d_var['is_kwonly'] = a.is_kwonly
+                        if isinstance(a, ValuedArgument):
+
+                            # optional argument only if the value is None
+                            if isinstance(a.value, Nil):
+                                d_var['is_optional'] = True
+
+                            a_new = ValuedFunctionAddress(a.name, ah.arguments, ah.results, [],
+                                        value=a.value, **d_var)
+                        else:
+                            a_new = FunctionAddress(a.name, ah.arguments, ah.results, [], **d_var)
+                    else:
+                        d_var = self._infere_type(ah, **settings)
+                        d_var['shape'] = ah.alloc_shape
+                        d_var['is_argument'] = True
+                        d_var['is_kwonly'] = a.is_kwonly
+                        dtype = d_var.pop('datatype')
+                        if d_var['rank']>0:
+                            d_var['cls_base'] = NumpyArrayClass
 
-                        # optional argument only if the value is None
-                        if isinstance(a.value, Nil):
-                            d_var['is_optional'] = True
+                        # this is needed for the static case
+                        if isinstance(a, ValuedArgument):
 
-                        a_new = ValuedVariable(dtype, str(a.name),
-                                    value=a.value, **d_var)
-                    else:
-                        a_new = Variable(dtype, a.name, **d_var)
+                            # optional argument only if the value is None
+                            if isinstance(a.value, Nil):
+                                d_var['is_optional'] = True
+
+                            a_new = ValuedVariable(dtype, str(a.name),
+                                        value=a.value, **d_var)
+                        else:
+                            a_new = Variable(dtype, a.name, **d_var)
 
                     if additional_args:
                         args += additional_args
 
                     args.append(a_new)
-                    self.insert_variable(a_new, name=str(a_new.name))
-
+                    if isinstance(a_new, FunctionAddress):
+                        self.insert_function(a_new)
+                    else:
+                        self.insert_variable(a_new, name=str(a_new.name))
```

Comment:
Is this required a lot? If so would it be worth modifying `insert_variable` instead so it starts with:
```
if isinstance(var, FunctionAddress):
    return self.insert_function(var)
```
?

Reply:
it's only used once, should I modify `insert_variable` however ?

Reply:
If you think it might be needed every time we use `insert_variable` then I would modify it. If this will be the only place it will ever be needed then leave it